### PR TITLE
Add helper decorator for JITing parametrized functions

### PR DESCRIPTION
**Pull request by eb8680**
Addresses #1063 

This PR adds a helpful utility `pyro.ops.jit.compile` that serves as a drop-in replacement for `torch.jit.compile` that is compatible with functions that call `pyro.param` internally.  Every such function is implicitly a closure over the Pyro parameter store, but `torch.jit.compile` does not allow closures over tensors.  `pyro.ops.jit.compile` enables this use case without passing the entire parameter store as an argument and also ensures that parameters are always exposed, even after compilation.

This makes the JIT easier to use for common Pyro use cases like computing loss functions in SVI or the energy in HMC that depend on parameters:
```python
def model(x):
    scale = pyro.param("scale", torch.tensor(0.5), constraint=constraints.positive)
    return pyro.sample("y", dist.Normal(x, scale))

@pyro.ops.jit.compile(nderivs=1)
def model_log_prob_fn(x, y):
    cond_model = pyro.condition(model, data={"y": y})
    tr = pyro.poutine.trace(cond_model).get_trace(x)
    return tr.log_prob_sum()

...

# even after compilation, model_log_prob_fn still exposes params
with poutine.trace(param_only=True) as param_capture:
    model_log_prob_fn(torch.tensor(1.0), torch.tensor(2.0))

assert "p" in param_capture.trace
```

The PR also slightly refactors `JitTrace*_ELBO` to use `pyro.ops.jit.compile` instead of `torch.jit.compile` and modifies `ReplayMessenger` to support replaying of parameters.

### CODE REVIEW

File: pyro/infer/trace_elbo.py [hunk 1]
```diff
@@ -166,18 +165,11 @@ class JitTrace_ELBO(Trace_ELBO):
...
-                        pass
-            self._param_names = list(param_capture.trace.nodes.keys())
-
             # build a closure for loss_and_surrogate_loss
             weakself = weakref.ref(self)
 
-            @torch.jit.compile(nderivs=1)
-            def loss_and_surrogate_loss(args_list, param_list):
+            @pyro.ops.jit.compile(nderivs=1)
+            def loss_and_surrogate_loss(*args_list):
```
Comment (fritzo) [inline pyro/infer/trace_elbo.py]: nit: I guess we can just call this `*args` now to match `Trace_ELBO.loss_and_grads()` idioms?

**Comment by eb8680** [comment]
GitHub's weird new merging behavior cluttered up the commit history...

File: pyro/infer/traceenum_elbo.py [hunk 2]
```diff
@@ -171,30 +170,22 @@ class JitTraceEnum_ELBO(TraceEnum_ELBO):
...
             weakself = weakref.ref(self)
 
-            @torch.jit.compile(nderivs=1)
-            def differentiable_loss(args_list, param_list):
+            @pyro.ops.jit.compile(nderivs=1)
+            def differentiable_loss(*args_list):
                 self = weakself()
                 elbo = 0.0
-                for model_trace, guide_trace in self._get_traces(model, guide, *args_list, **kwargs):
+                for model_trace, guide_trace in self._get_traces(model, guide,
```
Comment (fritzo) [inline pyro/infer/traceenum_elbo.py]: nit: Can we replace this with a single line to match other uses in the file:
```py
for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
```

File: pyro/infer/tracegraph_elbo.py [hunk 3]
```diff
@@ -303,18 +304,11 @@ class JitTraceGraph_ELBO(TraceGraph_ELBO):
...
-                        pass
-            self._param_names = list(param_capture.trace.nodes.keys())
-
             # build a closure for loss_and_surrogate_loss
             weakself = weakref.ref(self)
 
-            @torch.jit.compile(nderivs=1)
-            def loss_and_surrogate_loss(args_list, param_list):
+            @pyro.ops.jit.compile(nderivs=1)
+            def loss_and_surrogate_loss(*args_list):
```
Comment (fritzo) [inline pyro/infer/tracegraph_elbo.py]: nit: ditto `*args`

File: pyro/ops/jit.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    Example::
+
+        def model(x):
+            scale = pyro.param("scale", torch.tensor(0.5), constraint=constraints.positive)
+            return pyro.sample("y", dist.Normal(x, scale))
+
+        @pyro.ops.jit.compile(nderivs=1)
+        def model_log_prob_fn(x, y):
+            cond_model = pyro.condition(model, data={"y": y})
```
Comment (fritzo) [inline pyro/ops/jit.py]: :+1: that's awesome we support fancy poutine usage!

File: pyro/ops/jit.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                        fn(*args, **kwargs)
+
+                self._param_names = list(set(first_param_capture.trace.nodes.keys()))
+
+                weakself = weakref.ref(self)
+
+                @torch.jit.compile(**self._kwargs)
+                def compiled(unconstrained_params, *args):
+                    self = weakself()
+                    constrained_params = {name: pyro.param(name, unconstrained_param)
```
Comment (fritzo) [inline pyro/ops/jit.py]: This isn't doing what it looks like it's doing, and it would be wrong if the unconstrained param were ever used to initialized the param. I think it would be more transparent to
```py
constrained_params = {}
for name, unconstrained_param in zip(self._param_names, unconstrained_params):
    constrained_param = pyro.param(name)  # assume param has been initialized
    assert constrained_param.unconstrained() is unconstrained_param
    constrained_params[name] = constrained_param
```

File: pyro/ops/jit.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+                    return poutine.replay(
+                        fn, params=constrained_params)(*args, **kwargs)
+
+                self.compiled = compiled
+
+            param_list = [pyro.param(name).unconstrained()
+                          for name in self._param_names]
+
+            with poutine.block(hide=self._param_names):
```
Comment (fritzo) [inline pyro/ops/jit.py]: Can you help me undersand why you're blocking and tracing here?

File: pyro/ops/jit.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                with poutine.trace(param_only=True) as param_capture:
+                    ret = self.compiled(param_list, *args, **kwargs)
+
+            new_params = filter(lambda name: name not in self._param_names,
```
Comment (fritzo) [inline pyro/ops/jit.py]: In what circumstances would we see new params? Should we handle them gracefully or error?

Comment (eb8680) [reply pyro/ops/jit.py]: If a new input argument shape configuration triggers recompilation and the resulting new trace has different parameters, we'll need to capture those.

Since `torch.jit.compile` handles this, we should too.

File: pyro/ops/jit.py (hunk 6 above)
Comment (eb8680) [reply pyro/ops/jit.py]: I guess I wanted to avoid having `pyro.param` be called twice for params in `self._param_names`, but I'm not sure that matters?

File: pyro/ops/jit.py (hunk 7 above)
Comment (fritzo) [reply pyro/ops/jit.py]: Does that mean we should also reset the compiled function, since its signature has changed?
```py
if any(new_params):
    self.compiled = None
```
or will the original version automatically work when different params are passed in?

File: pyro/ops/jit.py (hunk 6 above)
Comment (fritzo) [reply pyro/ops/jit.py]: Ok, I see that it makes sense to capture new params, and if you're capturing new params, it makes sense to block. Thanks for explaining.

File: pyro/poutine/replay_messenger.py [hunk 8]
```diff
@@ -8,42 +8,17 @@ class ReplayMessenger(Messenger):
...
+    def __init__(self, trace=None, params=None):
         """
         :param trace: a trace whose values should be reused
 
         Constructor.
         Stores trace in an attribute.
         """
         super(ReplayMessenger, self).__init__()
-        assert trace is not None, "must provide trace"
+        assert trace is not None or params is not None, "must provide trace"
```
Comment (fritzo) [inline pyro/poutine/replay_messenger.py]: nit: "must provide either trace or params"

File: pyro/ops/jit.py (hunk 7 above)
Comment (eb8680) [reply pyro/ops/jit.py]: The function's signature shouldn't change, since the parameters are always passed in through a list.  I'd prefer to delegate all bookkeeping to `torch.jit.compile`.

File: pyro/poutine/replay_messenger.py [hunk 9]
```diff
@@ -57,19 +32,23 @@ def _pyro_sample(self, msg):
...
-                    self.trace.nodes[g_name]["is_observed"]:
-                raise RuntimeError("site {} must be sample in trace".format(g_name))
+                return None
+            if guide_msg["type"] != "sample" or \
+                    guide_msg["is_observed"]:
+                raise RuntimeError("site {} must be sample in trace".format(name))
             msg["done"] = True
-            msg["value"] = self.trace.nodes[g_name]["value"]
+            msg["value"] = guide_msg["value"]
+            msg["infer"] = guide_msg["infer"]
```
Comment (fritzo) [inline pyro/poutine/replay_messenger.py]: Do I understand correctly that we never needed this before since we only replayed models against guides, but now we're replaying guides against guides as well? Or are you merely cleaning up?

File: tests/poutine/test_poutines.py [hunk 10]
```diff
@@ -98,19 +98,6 @@ def test_replay_full(self):
             assert_equal(model_trace.nodes[name]["value"],
                          guide_trace.nodes[name]["value"])
 
-    def test_replay_partial(self):
```
Comment (fritzo) [inline tests/poutine/test_poutines.py]: Did this break?

File: pyro/poutine/replay_messenger.py (hunk 8 above)
Comment (fritzo) [reply pyro/poutine/replay_messenger.py]: Actually since this is user-facing we should
```py
if trace is None and params is None:
    raise ValueError("must provide either trace or params")
```

File: tests/poutine/test_poutines.py (hunk 10 above)
Comment (eb8680) [reply tests/poutine/test_poutines.py]: No, it was a test for a dead code path

File: pyro/poutine/replay_messenger.py (hunk 9 above)
Comment (eb8680) [reply pyro/poutine/replay_messenger.py]: Just cleanup, the sites logic never got used

File: pyro/ops/jit.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            return pyro.sample("y", dist.Normal(x, scale))
+
+        @pyro.ops.jit.compile(nderivs=1)
+        def model_log_prob_fn(x, y):
+            cond_model = pyro.condition(model, data={"y": y})
+            tr = pyro.poutine.trace(cond_model).get_trace(x)
+            return tr.log_prob_sum()
+    """
+    def __init__(self, **kwargs):
+        self._kwargs = kwargs
```
Comment (fritzo) [inline pyro/ops/jit.py]: nit: Can we call this `self._jit_options` or something more evocative?

File: pyro/ops/jit.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        self.compiled = None
+        self._param_names = None
+
+    def __call__(self, fn):
+
+        assert self.compiled is None, \
+            "Cannot use a compile instance on more than one function"
+
+        def wrapper(*args, **kwargs):
```
Comment (fritzo) [inline pyro/ops/jit.py]: What do you think about adding a check like
```py
if self._fixed_kwargs is None:
    self._fixed_kwargs = kwargs
elif kwargs != self._fixed_kwargs:
    raise NotImplementedError('pyro.jit.compile does not support dynamic **kwargs')
```

Comment (eb8680) [reply pyro/ops/jit.py]: I don't know if that's true about `torch.jit.compile` though?

Comment (fritzo) [reply pyro/ops/jit.py]: ok

File: pyro/ops/jit.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            scale = pyro.param("scale", torch.tensor(0.5), constraint=constraints.positive)
+            return pyro.sample("y", dist.Normal(x, scale))
+
+        @pyro.ops.jit.compile(nderivs=1)
+        def model_log_prob_fn(x, y):
+            cond_model = pyro.condition(model, data={"y": y})
+            tr = pyro.poutine.trace(cond_model).get_trace(x)
+            return tr.log_prob_sum()
+    """
+    class wrapper(object):
```
Comment (fritzo) [inline pyro/ops/jit.py]: I would recommend defining this big class outside of `compile`:
```py
class CompiledFunction(object):  # formerly class wrapper
    def __init__(self, fn, jit_options):
        ...
```
and then adding decorator magic in the `compile` helper, something like
```
def compile(fn=None, **jit_options):
    if fn is None:
        return lambda fn: compile(fn, **jit_options)
    return CompiledFunction(fn, jit_options) 
```

**Comment by eb8680** [comment]
> Do you want to add any tests in this PR, or should we wait till next week to add tests of e.g. dynamic len(args) and new params?

We can add tests separately
